llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp

```diff
@@ -2572,14 +2572,23 @@ bool TargetLowering::SimplifyDemandedBits(
   }
   case ISD::TRUNCATE: {
     SDValue Src = Op.getOperand(0);
+    SDNodeFlags Flags = Op->getFlags();
 
     // Simplify the input, using demanded bit information, and compute the known
     // zero/one bits live out.
     unsigned OperandBitWidth = Src.getScalarValueSizeInBits();
     APInt TruncMask = DemandedBits.zext(OperandBitWidth);
     if (SimplifyDemandedBits(Src, TruncMask, DemandedElts, Known, TLO,
-                             Depth + 1))
+                             Depth + 1)) {
+      if (Flags.hasNoSignedWrap() || Flags.hasNoUnsignedWrap()) {
+        // Disable the nsw and nuw flags. We can no longer guarantee that we
+        // won't wrap after simplification.
+        Flags.setNoSignedWrap(false);
+        Flags.setNoUnsignedWrap(false);
+        Op->setFlags(Flags);
```

Comment:
The API for SDNodeFlags is much worse than FastMathFlags. Can we get a wrapper with bit operations so you can just do a ~WrapFlags?